### ISSUE TITLE
Feature lockdown for Textpattern 4.7 beta deadline

### ISSUE THREAD

**Issue by philwareham** [textpattern/textpattern#1041]
Hi fellow devs. The aim is to get a beta release of Textpattern out by Christmas - I'm proposing 20th December. That gives us just under 4 weeks to prepare.

Therefore new features, unless critical to 4.7 functionality, should probably be held back until after beta. I'm not opposed to adding small additional features during the beta period, and having some known issues also dealt with during that time. But large feature additions during the beta will have to be deferred until the 4.8 dev cycle.

The good news is that we plan to make the release cycles quicker after 4.7 by having fewer big features released per version, but releasing quicker. Therefore new features are still developed at the same rate but we get them into the hands of users in a more timely fashion and keep driving forward the momentum that has built over the last few months thanks to all your great work.

The theme feature should land in dev in a week or so, FYI.

I guess we will have at least a couple of betas before final 4.7.0 release so realistically it would be end of January at the earliest for that stable release.

Hope that works for everyone. Let me know if you have any concerns.

**Comment by NicolasGraph** [comment]
> The theme feature should land in dev in a week or so, FYI.

Sorry I'm late, but I'll spend most of the week on the this feature, adapting the code structure to make it usable by the _Pages_, _Forms_ and _Styles_ panels in the future.
See also https://github.com/textpattern/textpattern/issues/1035#issuecomment-349439389

**Comment by petecooper** [comment]
Can we nail down a date for beta-1, do you think? Textpattern's infamous feature creep is rearing its head again.

**Comment by philwareham** [comment]
Hi @petecooper - as far as I know as soon as I've completed the languages update on Crowdin and Stef has merged the languages branch, the beta can go ahead. I'm trying to squeeze them in as quickly as I can between my real paid work - about 13 languages left to fix (and it takes about 15 minutes per language for me to fix). If someone has other bits that need to be done before beta I'm unaware of them, but perhaps they can let us know. CC @Bloke @bloatware @makss

**Comment by petecooper** [comment]
Nice one @philwareham - thank you.

Anything you can defer to me gruntwork-wise?

**Comment by philwareham** [comment]
Not really, but thanks anyway. I just need to slog through them using the process I have in place. Basically, regrouping of language strings forced a big rewrite on Crowdin, which I hadn't foreseen.

**Comment by bloatware** [comment]
Have not we decided to remove `themes` from `setup` directory? It then needs renaming `default` to something else (e.g. `factory` or `hive`).

No new features planned on my side.

**Comment by Bloke** [comment]
Yeah, my fault about the language rejig, sorry. But it's worth it as the strings really were a mess in terms of grouping, and this change to the way of loading strings really highlighted the terrible organisational state of the language files. At least now we'll have a neater platform upon which to add strings, with better (and kind-of-enforced) content grouping.

In terms of creep, most of the stuff I've been doing is mopup. Off the top of my head, things left to do:

* Finalize the theme callbacks and ensure they're workable without needing to raise a whole tonne of them to do simple things. Testing required from any budding plugin authors who want to take themes up a notch already.
* Thoroughly test the language strings with plugins running from cache directory. I suspect we need to tweak the plugin loader here to read in the textpack from the file and merge it with the strings in the DB to simulate the plugin having been "installed". When new languages are installed, the cache dir needs combing too and any plugins that have a textpack need to be installed.
* Add new classes to checksums. Laborious. Can we automate this with a dev tool, @makss ?
* Sync up tag builder with reality / new attributes. I have _no idea_ how to handle global attributes in the tag builder so I'm pretending they don't exist ;-) @bloatware if you have any suggestions, I'd love to hear them. It's not a huge issue if the tag builder isn't 100% complete as it's a relatively minor part of the system and can be deferred to a point release as a 'bug fix' or when the -rc drops.
* Possibly merge in Nicolas' static-reduced rehash of the themes. Again, if it's not done, no harm done as it works now, this is just for efficiency reasons and can be bundled in a point release or for -rc.

Anything else?

**Comment by philwareham** [comment]
> Have not we decided to remove themes from setup directory? It then needs renaming default to something else (e.g. factory or hive).

How about `example` as the theme name?

**Comment by bloatware** [comment]
> How about *example* as the theme name?

@philwareham your theme - your name :-) But `example` sounds basic and incomplete, I'd prefer something more solid... `txp_4.7.0`?

**Comment by bloatware** [comment]
> I have no idea how to handle global attributes in the tag builder

@Bloke perhaps, append them globally, on the tagbuilder level? For some tags (e.g. `txp_die`) they are nonsense, but it's a minor annoyance. I'll take a look.

**Comment by philwareham** [comment]
Closing. Superseded by https://github.com/textpattern/textpattern/issues/1098.
